### PR TITLE
MULE-18479: Verbose/Wirelog log for TCP Connections

### PR DESCRIPTION
**Pull request by julodiaz**

### CODE REVIEW

File: src/main/java/org/mule/extension/socket/api/connection/tcp/protocol/XmlMessageProtocol.java [hunk 1]
```diff
@@ -99,6 +102,13 @@ public InputStream read(InputStream socketIs) throws IOException {
         message.setLength(patternIndex);
       }
 
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("Logging TCP Content:");
```
Comment (nicomz) [inline src/main/java/org/mule/extension/socket/api/connection/tcp/protocol/XmlMessageProtocol.java:106]: Please look at SocketUtils.java comments.

File: src/main/java/org/mule/extension/socket/api/connection/tcp/protocol/LengthProtocol.java [hunk 2]
```diff
@@ -118,6 +129,13 @@ public void write(OutputStream os, InputStream inputStream) throws IOException {
     dataOutputStream.writeInt(data.length);
     dataOutputStream.write(data);
 
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("Logging TCP Content:");
+      LOGGER.debug("-----------------------------------");
+      LOGGER.debug(Base64.getEncoder().encodeToString(data));
+      LOGGER.debug("-----------------------------------");
+    }
+
```
Comment (nicomz) [inline src/main/java/org/mule/extension/socket/api/connection/tcp/protocol/LengthProtocol.java:138]: Please look at SocketUtils.java comments.
Static Import of getEncoder

File: src/main/java/org/mule/extension/socket/api/connection/tcp/protocol/LengthProtocol.java [hunk 3]
```diff
@@ -95,6 +99,13 @@ public InputStream read(InputStream socketIs) throws IOException {
     byte[] buffer = new byte[length];
     dis.readFully(buffer);
 
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("Logging TCP Content:");
+      LOGGER.debug("-----------------------------------");
+      LOGGER.debug(Base64.getEncoder().encodeToString(buffer));
+      LOGGER.debug("-----------------------------------");
+    }
+
```
Comment (nicomz) [inline src/main/java/org/mule/extension/socket/api/connection/tcp/protocol/LengthProtocol.java:108]: Please look at SocketUtils.java comments

File: src/main/java/org/mule/extension/socket/api/connection/tcp/protocol/DirectProtocol.java [hunk 4]
```diff
@@ -68,6 +70,13 @@ public InputStream read(InputStream socketIs) throws IOException {
     } while (repeat);
     byteArrayOutputStream.flush();
 
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("Logging TCP Content (Base64 encoding):");
+      LOGGER.debug("-----------------------------------");
+      LOGGER.debug(Base64.getEncoder().encodeToString(byteArrayOutputStream.toByteArray()));
+      LOGGER.debug("-----------------------------------");
+    }
+
```
Comment (nicomz) [inline src/main/java/org/mule/extension/socket/api/connection/tcp/protocol/DirectProtocol.java:79]: Please look at SocketUtils.java class comments.
Static import of getEncoder

File: src/main/java/org/mule/extension/socket/api/connection/tcp/protocol/DirectProtocol.java [hunk 5]
```diff
@@ -70,12 +70,7 @@ public InputStream read(InputStream socketIs) throws IOException {
     } while (repeat);
     byteArrayOutputStream.flush();
 
-    if (LOGGER.isDebugEnabled()) {
-      LOGGER.debug("Logging TCP Content (Base64 encoding):");
-      LOGGER.debug("-----------------------------------");
-      LOGGER.debug(Base64.getEncoder().encodeToString(byteArrayOutputStream.toByteArray()));
-      LOGGER.debug("-----------------------------------");
-    }
+    logIfDebugEnabled(byteArrayOutputStream.toByteArray(), LOGGER);
```
Comment (nicomz) [inline src/main/java/org/mule/extension/socket/api/connection/tcp/protocol/DirectProtocol.java:73]: You're reading the byteArrayOutputStream even if debug is disabled.

File: src/main/java/org/mule/extension/socket/internal/SocketUtils.java [hunk 6]
```diff
@@ -174,17 +175,28 @@ public static void sendUdpPackages(InputStream inputStream, int bufferSize, Sock
...
-  public static InputStream logIfNeeded(InputStream content, Logger logger) {
+  public static InputStream logIfDebugEnabled(InputStream content, Logger logger) {
     if (logger.isDebugEnabled()) {
-      String c = IOUtils.toString(content);
-      logger.debug("Logging TCP Content (Base64 encoding):");
-      logger.debug("-----------------------------------");
-      logger.debug(Base64.getEncoder().encodeToString(c.getBytes()));
-      logger.debug("-----------------------------------");
-      return new ByteArrayInputStream(c.getBytes());
+      byte[] bytes = IOUtils.toByteArray(content);
```
Comment (nicomz) [inline src/main/java/org/mule/extension/socket/internal/SocketUtils.java:183]: I don't recommend doing this. You might be loading a lot of data into memory. Read and Log chunks of the input stream. There's some challenge having to return an InputStream.

**Comment by julodiaz** [comment]
We can safely use InputStream.available() since we use the BufferedInputStream implementation.

File: src/test/java/org/mule/extension/socket/TcpLoggerForInputStreamTestCase.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  public void testLogIfDebugEnabled() throws Exception {
+    InputStream inputStream = Mockito.mock(InputStream.class);
+    Logger logger = Mockito.mock(Logger.class);
+
+    Mockito.when(logger.isDebugEnabled()).thenReturn(true);
+    SocketUtils.logIfDebugEnabled(inputStream, logger);
+
+    Mockito.verify(logger, Mockito.atLeastOnce()).debug(Mockito.anyString(), Mockito.anyString(), Mockito.anyString(),
+                                                        Mockito.anyString());
+  }
```
Comment (nicomz) [inline src/test/java/org/mule/extension/socket/TcpLoggerForInputStreamTestCase.java:27]: The goal of LogIfDebugEnabled is to only log when debug is enabled for current class. This is the happy path test. 
What happens when the debug is disabled? The behavior of not reading the inputStream should be tested. Also the overloads of the method should be tested.

**Pull request merged**